Implement evaluators for checkcast and checkcastAndNULLCHK

aarongraham9 opened the pull request:
- Implement evaluators for checkcast and checkcastAndNULLCHK for AArch64
- Implement TEMPORARY_initJ9ARM64TreeEvaluatorTable for Aarch64

Signed-off-by: Aaron Graham <aaron.graham@unb.ca>

aarongraham9 commented:
After CI passes this should be good for a review @0xdaryl.

0xdaryl commented:
All that is really required for these kinds of check nodes right now is to simply (temporarily) convert them to a call, evaluate them, and then switch the opcode back.  This essentially punts to the VM to handle, which will be the eventual fallback path when these nodes are optimized.

Eventually what you're implementing here is the start of an optimized version of those opcodes, but I don't think we need that right now (and frankly, I don't want to sort through all the subtleties of the optimization at this point).  Please keep it in your private fork for now until we want to optimize these opcodes.

So all you need right now to implement this is based largely on the code you've already written here:
```
      TR::ILOpCodes opCode = node->getOpCodeValue();
      TR::Node::recreate(node, TR::call);
      TR::Register *targetRegister = directCallEvaluator(node, cg);
      TR::Node::recreate(node, opCode);
```

aarongraham9 commented:
Ok. I'll update the PR accordingly. Thanks!

aarongraham9 commented:
Rebased off of `master` and cut back the evaluator to just do the call  and punt back to the VM. Once CI finishes this is good to go.

aarongraham9 commented:
Rabsed off of `master`, fixed up the include and added in the `TEMPORARY_initJ9ARM64TreeEvaluatorTable` override table. Local build and CI passed; this is good to go @0xdaryl.

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -20,8 +20,61 @@
+
+   tet[TR::tstart] = TR::TreeEvaluator::tstartEvaluator;
+   tet[TR::tfinish] = TR::TreeEvaluator::tfinishEvaluator;
+   tet[TR::tabort] = TR::TreeEvaluator::tabortEvaluator;
```
Leave the transactional memory evaluators as unimplemented for now.  Mainly because AArch64 support for TM was only just announced (i.e., not available yet) and also because I can't find an implementation of these functions.

aarongraham9 replied:
Ok will do!

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -20,8 +20,61 @@
+   tet[TR::dcall]  = TR::TreeEvaluator::directCallEvaluator;
+   tet[TR::acall]  = TR::TreeEvaluator::directCallEvaluator;
+   tet[TR::call]   = TR::TreeEvaluator::directCallEvaluator;
+   tet[TR::vcall]  = TR::TreeEvaluator::directCallEvaluator;
```
Why are these being initialized here?  This is the default value that the OMR code generator initializes them to.

aarongraham9 replied:
Not sure Just following the lead of the P compiler for this one.

0xdaryl commented:
Also, please change your commit title to something like `Implement evaluators for checkcast and checkcastAndNULLCHK`

aarongraham9 commented:
I'll have a new version for you shortly.

aarongraham9 commented:
Rabsed off of master, fixed up the merge conflicts and separated the PR into two commits @0xdaryl.

aarongraham9 commented:
CI finished. This is good to go @0xdaryl.

0xdaryl commented:
The code changes look fine.  Minor change required:  please remove the triplicate text from each commit message.  Only one of the lines will do.  Thanks.

aarongraham9 commented:
Ok, done! Rebased off of `master` and fixed up the merge conflicts arising from the merging of the `DIVCHK` Evaluator in #6055. As this added the `TEMPORARY_initJ9ARM64TreeEvaluatorTable` I've enabled the `DIVCHK` J9 Evaluator in the `TEMPORARY_initJ9ARM64TreeEvaluatorTable`.

0xdaryl commented:
This is an AArch64-specific change.  It is pointless to wait for the CI build.
